### DISCUSSION TITLE
Get default value on initial render

### DISCUSSION THREAD

**Discussion by zyasir** [react-hook-form/react-hook-form#5190] [Q&A] [answered]
I am using v7. I am trying to get the default value on initial render, but it's causing me unexpected results.

I have this parent componnt:
```
const form = useForm({ defaultValues: { foo: { bar: 45 } } });
<FormProvider {...form}>
```

Then this is my component:
```
function InputRow(props) {

  const name = 'foo.bar';
  const form = useFormContext();

  const [isZero, setIsZero] = useState(() => {
    // useState callback as this runs only on mount
   const mountValue = form.getValues(name);

   if (mountValue === 0) {
     return true;
   } else {
     return false;
   }
  });

  return (
    <HookedTextInput name={name} />
  );
}
```

When I do this, `getValues` is coming out to be `undefined`, I expect `45`.

The docs say don't use `form.getValues` in render, so then I tried to use `const value = form.watch(name)` but this also gave `undefined`.

Anyway to get default values in initial render?

### Also

Is the docs wrong here in this example - https://react-hook-form.com/api/useform/getvalues

<img width="740" alt="image" src="https://user-images.githubusercontent.com/49174024/118208890-674b3c80-b41c-11eb-9db3-7230b4c3bdf6.png">

The last 2 examples are:

```

getValues("yourDetails.firstName") | { lastName: '' }
-- | --
getValues(["yourDetails.lastName"]) | [{ firstName: '' }]

```

Should this be changed to the following?

```


getValues("yourDetails.firstName") | { firstName: '' }
-- | --
getValues(["yourDetails.lastName"]) | [{ lastName: '' }]

```

**Comment by zyasir** [comment]
Very interesting. I downgraded to v6 and it's working.

**Reply by zyasir** [reply]
I think it's a mix of having to do with keeping the inputs rendered. Initial render it takes from defaultValues, but then it seems the field is not mounted, so it then overwrites the value to be undefined. And when it overrides the value, it doesn't trigger the `form.watch(callback)` callback I have. Something interesting going on here. Even in v6, it didnt work as expected until I kept the field rendered by hidden with `display: none`

**Comment by leapful** [comment, marked as answer]
@zyasir 

react-hook-form depends on ref of form element in order to initialize form data, if the form is not rendered yet, form handling is not ready therefore you cannot get value at that phase.

I would suggest to use another Context to get default values instead of getting it from react-hook-form, something like:

```

const FormDefaultValuesContext = createContext();

// Wrap it with FormProvider from react-hook-form

const form = useForm({ defaultValues });

<FormDefaultValuesContext.Provider value={defaultValues}>
  <FormProvider {...form}>
      {children}
  </FormProvider>
</FormDefaultValuesContext.Provider>

// then use it inside form element component

const InputRow = () => {
   const defaultValues = useContext(FormDefaultValueContext);
   // Handling initial state with defaultValues[name]
}
```

**Reply by zyasir** [reply]
Very cool, this makes sense with the nature of uncontrolled.

**Reply by bluebill1049** [reply]
thank you very much @leapful 🙏
